RELEASE_NOTES.md

```diff
@@ -16,6 +16,34 @@
 
 - HTTP2 keep-alive is now enabled by default, with an interval of 60 seconds between pings, and a 20 second timeout for responses from the service.  These values are configurable and may be updated based on specific requirements.
 
+* The `BaseApiClient` class is not generic anymore, and doesn't take a function to create the stub. Instead, subclasses should create their own stub right after calling the parent constructor. This enables subclasses to cast the stub to the generated `XxxAsyncStub` class, which have proper `async` type hints. To convert you client:
+
+    ```python
+    # Old
+    from my_service_pb2_grpc import MyServiceStub
+    class MyApiClient(BaseApiClient[MyServiceStub]):
+        def __init__(self, server_url: str, *, ...) -> None:
+            super().__init__(server_url, MyServiceStub, ...)
+            ...
+
+    # New
+    from typing import cast
+    from my_service_pb2_grpc import MyServiceStub, MyServiceAsyncStub
+    class MyApiClient(BaseApiClient):
+        def __init__(self, server_url: str, *, ...) -> None:
+            super().__init__(server_url, connect=connect)
+            self._stub = cast(MyServiceAsyncStub, MyServiceStub(self.channel))
+            ...
```

Comment:
Can't we just ask them to pass `cast(MyServiceAsyncStub, MyServiceStub(service.channel))` as an argument, and keep make the base client generic over `MyServiceAsyncStub` instead?

Reply:
No, because `__init__` can have `MyServiceAsyncStub` as a type hint, to do that we need to take the type as a generic, and we can't use an async stub as a generic because it is parsed by the interpreter, and it can't find it because it is not present in the `.py` file, only the `.pyi` file.

This is how great grpio is handling asyncio and type-hinting :grimacing: (there is a new experimental interface in the cooking, let's see how that turns out).

Reply:
It won't parse them with `from __future__ import annotations` right?

Reply:
never mind, it will parse them

Reply:
Yes, because it is not an annotation in this case, is is a proper object for the interpreter. You can give it a go if you want, I already tried many different attempts and nothing seemed to work.

```console
$ head -n 100 t.py* t2.py 
==> t.py <==
class IDoExist:
    pass

==> t.pyi <==
class IDoExist:
    pass

class IDontExist:
    pass

==> t2.py <==
from __future__ import annotations

from typing import Generic, TypeVar

from t import IDoExist, IDontExist

T = TypeVar("T")


class G(Generic[T]):
    def __init__(self, x: T):
        self.x = x


class Sub(G[IDoExist]):
    pass


class Sub2(G[IDontExist]):
    pass
$ python t2.py 
Traceback (most recent call last):
  File "/home/luca/devel/client-base/t2.py", line 5, in <module>
    from t import IDoExist, IDontExist
ImportError: cannot import name 'IDontExist' from 't' (/home/luca/devel/client-base/t.py)
```

Reply:
Put another way, with `TYPE_CHECKING`:

```py
from t import IDoExist

if TYPE_CHECKING:
    from t import IDontExist
```

Then the error becomes:

```console
$ python t2.py 
Traceback (most recent call last):
  File "/home/luca/devel/client-base/t2.py", line 22, in <module>
    class Sub2(G[IDontExist]):
                 ^^^^^^^^^^
NameError: name 'IDontExist' is not defined. Did you mean: 'IDoExist'?
```

Maybe it is more clear to see this way. Only stuff after a `:` (or used in special constructs, like `cast`, are type hints.